### PR TITLE
NDS, stripping out all the other matching stuff

### PR DESCRIPTION
**Pull request by deanWombourne**
Hey. The is a PR implementing #20 and stripping out all the other matching / comparable stuff from `StylistIdentifier`.

It's only a draft because 

(a) I'm still not 100% sure I've thought of all the edge cases where we want this to match/no match

(b) I haven't even begun to think about matching variants, just identifiers.

### CODE REVIEW

File: StylableSwiftUI/Classes/StylableImage.swift [hunk 1]
```diff
@@ -90,8 +90,9 @@ extension StylistIdentifier {
                              separator: String = StylableImage.defaultSeparator,
                              maxLength: Int = StylableImage.defaultMaxLength) -> AnySequence<String> {
 
-        let components = self.components.reversed()
-        let options = VariantSequence(from: Array(components), wildcard: wildcard, maxLength: maxLength)
+        var components = Array(self.path.components.reversed())
+        components.append(Component(value: self.token, state: nil))
+        let options = VariantSequence(from: components, wildcard: wildcard, maxLength: maxLength)
```
Comment (kerrmarin) [inline StylableSwiftUI/Classes/StylableImage.swift:95]: Do we need wildcards anymore?

File: StylableSwiftUI/Classes/StylistIdentifier.swift [hunk 2]
```diff
@@ -137,46 +73,64 @@ extension StylistIdentifier: ExpressibleByStringLiteral {
...
-// MARK: - Component
+        public var description: String {
+            self.components.reversed().map(String.init).joined(separator: "/")
+        }
 
-extension StylistIdentifier {
+        var isEmpty: Bool { self.components.isEmpty }
+
+        func component(at index: Int) -> Component? {
+            guard index < self.components.count else { return nil }
```
Comment (kerrmarin) [inline StylableSwiftUI/Classes/StylistIdentifier.swift]: And more than or eq to 0?

File: StylableSwiftUI/Classes/StylistIdentifier.swift [hunk 3]
```diff
@@ -137,46 +73,64 @@ extension StylistIdentifier: ExpressibleByStringLiteral {
...
+            return self.components[index]
+        }
+
+        func within(_ path: Path?) -> Path {
+            guard let path = path else { return self }
```
Comment (kerrmarin) [inline StylableSwiftUI/Classes/StylistIdentifier.swift]: what's the point of having an optional Path here if we're just gong to return self? Doesn't it make more sense to not allow nil values?

Comment (kerrmarin) [reply StylableSwiftUI/Classes/StylistIdentifier.swift]: And if we're optmizing, should we check the given path is not empty?

File: StylableSwiftUI/Classes/StylistIdentifier.swift [hunk 4]
```diff
@@ -137,46 +73,64 @@ extension StylistIdentifier: ExpressibleByStringLiteral {
...
     struct Component: CustomStringConvertible, Equatable, Hashable {
-        let value: String?
-        let state: String?
+        let value: String
+        let variant: String?
 
-        init(value: String?, state: String?) {
-            self.value = value != "*" ? value : nil
-            self.state = state
+        init(value: String, state: String?) {
```
Comment (kerrmarin) [inline StylableSwiftUI/Classes/StylistIdentifier.swift]: `value: String, variant: String?`?

File: StylableSwiftUI/Classes/StylistIdentifier.swift [hunk 5]
```diff
@@ -246,31 +170,21 @@ extension StylistIdentifier.Component: ExpressibleByStringLiteral {
...
 
-    /// Create an identifier representing `self` inside `identifier`
+    /// Create an identifier representing `self` inside `path`
     ///
     /// i.e. `"close".within("button") == 'button/close'`
     ///
-    public func within(_ identifier: StylistIdentifier?) -> StylistIdentifier {
-        guard let identifier = identifier else { return self }
+    public func within(_ path: Path?) -> StylistIdentifier {
+        guard let path = path else { return self }
```
Comment (kerrmarin) [inline StylableSwiftUI/Classes/StylistIdentifier.swift]: Don't think we need to do this since the (current) `within` method takes an optional

**Review by kerrmarin — APPROVED** [review]
LGTM -- I'm not sure that the image identifiers still need wildcards

File: StylableSwiftUI/Classes/StylableImage.swift (hunk 1 above)
Comment (deanWombourne) [reply StylableSwiftUI/Classes/StylableImage.swift:95]: I don't think so, but I couldn't work out how to make all the image name options to run through so I left it for another PR :)

File: StylableSwiftUI/Classes/StylistIdentifier.swift (hunk 3 above)
Comment (deanWombourne) [reply StylableSwiftUI/Classes/StylistIdentifier.swift]: (I think) It makes some of the callers lives easier - I'll check. And if we keep it then we should definitely check to see if it's empty.

File: StylableSwiftUI/Classes/StylistIdentifier.swift (hunk 4 above)
Comment (deanWombourne) [reply StylableSwiftUI/Classes/StylistIdentifier.swift]: Sigh, yep.

File: StylableSwiftUI/Classes/StylistIdentifier.swift (hunk 3 above)
Comment (deanWombourne) [reply StylableSwiftUI/Classes/StylistIdentifier.swift]: Only used in one place - so I removed the optionality.

File: StylableSwiftUI/Classes/StylistIdentifier.swift (hunk 2 above)
Comment (deanWombourne) [reply StylableSwiftUI/Classes/StylistIdentifier.swift]: Done

File: StylableSwiftUI/Classes/StylistIdentifier.swift (hunk 5 above)
Comment (deanWombourne) [reply StylableSwiftUI/Classes/StylistIdentifier.swift]: Hah, investigated this and completely changed my mind about the optional in `Path.within(_:)`! This is used in a few places and it's far more readable just to deal with optionals internally, so I've put your isEmpty check in and left them as optionals! 

Removed the path=path line from this method - it's only really an optimisation if StylistIdentifier is a class - but it's a struct so I don't think we win anything by "optimising" here.

File: StylableSwiftUI/Classes/StylistIdentifier.swift (hunk 3 above)
Comment (deanWombourne) [reply StylableSwiftUI/Classes/StylistIdentifier.swift]: ^ Found more places with use it, so put the optional back in and added your isEmpty optimisation check :)

**Comment by deanWombourne** [comment]
@kerrmarin OK, think I've fixed everything aside from the image name wildcard stuff, which I think needs an actual real chat about how it's going to work.

File: StylableSwiftUI/Classes/Logger.swift [hunk 6]
```diff
@@ -8,33 +8,77 @@
...
-
 public final class Logger {
 
+    // Logging level. Given the raw values of OSLogType I'm not sure I trust it enough.
+    public enum Level: Int {
+        case `default` = 0
+        case info = 1
+        case debug = 2
+        case error = 3
+        case fault = 4
```
Comment (atomoil) [inline StylableSwiftUI/Classes/Logger.swift:19]: fatal ?

**Review by atomoil — APPROVED** [review]

**Pull request merged**